openedx_tagging/core/tagging/api.py

```diff
@@ -98,5 +114,180 @@ def tag_object(
+        )
+
+    # Read file and build the tags data to be uploaded
+    try:
+        tags_data = {}
+        tags.seek(0)
+        if format == TaxonomyDataFormat.CSV:        
+            text_tags = TextIOWrapper(tags, encoding='utf-8')
+            csv_reader = csv.DictReader(text_tags)
+            header_fields = csv_reader.fieldnames
+            if csv_fields != header_fields:
+                raise ValueError(
+                    _(
+                        f"Invalid CSV header: {header_fields}. Must be: {csv_fields}."
+                    )
+                )
+            tags_data = list(csv_reader)
+        else:
+            # TaxonomyDataFormat.JSON
+            tags_data = json.load(tags)
+            if 'tags' not in tags_data:
+                raise ValueError(
+                    _(
+                        f"Invalid JSON format: Missing 'tags' list."
+                    )
+                )
+            tags_data = tags_data.get('tags')
+    except ValueError as e:
+        raise e
+    finally:
+        tags.close()
+
+
+    new_tags = []
+    updated_tags = []
+
+    def create_update_tag(tag):
+        """
+        Function to create a new Tag or update an existing one.
+
+        This function keeps a creation/update history with `new_tags` and `updated_tags`,
+        a same tag can't be created/updated in a same taxonomy import.
+        Also, recursively, creates the parents of the `tag`.
+
+        Returns the created/updated Tag.
+        Raise KeyError if 'id' or 'name' don't exist on `tag`
+        """
+
+        tag_id = tag['id']
+        tag_name = tag['name']
+        tag_parent_id = tag.get('parent_id')
+        tag_parent_name = tag.get('parent_name')
+
+        # Check if the tag has not already been created or updated
+        if tag_id not in new_tags and tag_id not in updated_tags:
+            try:
+                # Update tag
+                tag_instance = Tag.objects.get(external_id=tag_id)
+                tag_instance.value = tag_name
+
+                if tag_instance.parent and (not tag_parent_id or not tag_parent_name):
+                    # if there is no parent in the data import
+                    tag_instance.parent = None
+                updated_tags.append(tag_id)
+            except ObjectDoesNotExist:
+                # Create tag
+                tag_instance = Tag(
+                    taxonomy=taxonomy,
+                    value=tag_name,
+                    external_id=tag_id,    
+                )
+                new_tags.append(tag_id)
+
+            if tag_parent_id and tag_parent_name:
+                # Parent creation/update
+                parent = create_update_tag({'id': tag_parent_id, 'name': tag_parent_name})
+                tag_instance.parent = parent
+
+            tag_instance.save()
+            return tag_instance
+        else:
+            # Returns the created/updated tag from history
+            return Tag.objects.get(external_id=tag_id)
```

Comment:
Same thing here, needs `taxonomy=taxonomy`. The same `external_id` can appear in multiple taxonomies.

BTW if you add something like this to the `Taxonomy` class:

```python
@property 
def tags(self):
    return Tag.objects.filter(taxonomy=self)
```

then this code can become `return taxonomy.tags.get(external_id=tag_id)`

Reply:
There's already `taxonomy.tag_set`, so I don't think we need another property?

Reply:
Oh right, duh... not sure why I forgot that. But yes, use that :)